doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+header includes `Upgrade` in its list of values. If the `Upgrade` header exists, and its
+value is `h2c`, the connection is upgraded to HTTP/2 before continuing. If it has any
+other value, then the request should fail with a 500[1](#http-1).
+
+The proxy honors hop-by-hop headers (i.e. `Connection`, `Authorization`) and it removes
+them from requests before they are routed.
+
+The proxy honors `CONNECT` requests as any other. Once the request is routed to an
+appropriate destination, it is dispatched, and the accepted connection is treated as part
+of the request.
+
+All HTTP/1.1 requests must have a valid `Host` value.
+
+### HTTP/2
+
+The proxy honors `CONNECT` requests as any other.
+
+The proxy honors hop-by-hop headers (i.e. `Connection`, `Authorization`) and it removes
+them from requests before they are routed.
+
+All HTTP/2 requests must have a valid `:authority` value.
+
+### Routing
+
+HTTP/2 requests are routed by their `:authority`, and HTTP/1 requests by their `Host`.
+We'll refer to all of these as an _Authority_.
+
+#### Service Discovery
+
+In order to perform request-level load balancing, the proxy needs to resolve the Authority
+to a set of TCP addresses.
+
+The proxy issues a request to the Conduit control plane with the _Authority_, as specified
+in the request, as well as a set of _destination search paths_, specified to proxy at
+start-time. The control plane attempts to resolve the authority (using its search paths)
+through its known service discovery backends.
+
+For example, suppose we attempt to route a request with an authority of _rabbits_. If the
+proxy's running in a kubernetes cluster in the _farm_ namespace, the proxy should be
+configured with a destination search path like:
+```
+farm.svc.cluster.local.
+svc.cluster.local.
+```
+
+The control plane attempts to discover endpoints via the names:
+1. `rabbits.`
+2. `rabbits.farm.svc.cluster.local.`
+3. `rabbits.svc.cluster.local.`
+
+If the service actually exists, the control plane returns a result indicating that it
+bound to the FQDN `rabbits.farm.svc.cluster.local.` and provides a stream of updates for
+this service.
+
+If the controller cannot resolve a name to a set of addresses (e.g. because the service
+doesn't exist), it fails the resolution with a negative result indicating whether local
+DNS resolution may be attempted by the proxy. If so, the proxy performs DNS resolution
+locally (honoring /etc/nsswitch.conf, /etc/resolv.conf, etc). This is necessary because we
+cannot assume that the proxy and control plane have the same DNS configuration (and, in
+fact, we know that they don't in major cloud providers like GCP and AWS).
+
+### Questions
+
+<a name="http-q1"></a>
+1. I think that we need to _fail_ requests that include unsupportable upgrades. The
+   alternative would be to downgrade the connection to Opaque TCP, but then this becomes a
+   trivial vector to bypass policy--just add `Upgrade: NoPolicy!`. We'll need to rely on external policy to skip
```

Comment:
Policies should be whitelists, then this concern doesn't matter, AFAICT, since everything with `Upgrade: NoPolicy` should be rejected. Also, `Upgrade` is hop-by-hop so we can't really support unknown upgrades anyway.